### PR TITLE
Servantify cargohold and add `federationDomain` option

### PR DESCRIPTION
**Pull request by pcapriotti**
This converts the public API to Servant, removes the dependency on wai-routing, and also introduces a mandatory `federationDomain` configuration option.

Tracked by https://wearezeta.atlassian.net/browse/SQCORE-1174.

## Checklist

 - [x] The **PR Title** explains the impact of the change.
 - [x] The **PR description** provides context as to why the change should occur and what the code contributes to that effect. This could also be a link to a JIRA ticket or a Github issue, if there is one.
 - [x] If HTTP endpoint paths have been added or renamed, the **endpoint / config-flag checklist** (see Wire-employee only backend [wiki page](https://github.com/zinfra/backend-wiki/wiki/Checklists)) has been followed.
 - [x] If a cassandra schema migration has been added, I ran **`make git-add-cassandra-schema`** to update the cassandra schema documentation.
 - [x] **changelog.d** contains the following bits of information ([details](https://github.com/wireapp/wire-server/blob/develop/docs/developer/changelog.md)):
   - [x] A file with the changelog entry in one or more suitable sub-sections. The sub-sections are marked by directories inside `changelog.d`.
   - [x] If public end-points have been changed or added: does nginz need un upgrade?
   - [x] If internal end-points have been added or changed: which services have to be deployed in a specific order?

### CODE REVIEW

File: services/cargohold/src/CargoHold/Options.hs [hunk 1]
```diff
@@ -78,7 +79,20 @@ data Settings = Settings
...
+    -- Federation domain is used to qualify local IDs and handles,
+    -- e.g. 0c4d8944-70fa-480e-a8b7-9d929862d18c@wire.com and somehandle@wire.com.
+    -- It should also match the SRV DNS records under which other wire-server installations can find this backend:
+    --    _wire-server-federator._tcp.<federationDomain>
+    -- Once set, DO NOT change it: if you do, existing users may have a broken experience and/or stop working
+    -- Remember to keep it the same in Galley.
+    -- Example:
+    --   setFederationAllowedDomains:
+    --     - wire.com
+    --     - example.com
```
Comment (jschaul) [inline services/cargohold/src/CargoHold/Options.hs:94]:
```suggestion
    -- Remember to keep it the same in Galley and in Brig.
    -- This is referred to as the 'backend domain' in the public documentation; See
    -- https://docs.wire.com/how-to/install/configure-federation.html#choose-a-backend-domain-name 
```

**Review by jschaul — APPROVED** [review]
Approving the federationDomain setting option in its various incantations. The rest I just skimmed.

File: changelog.d/0-release-notes/cargohold-domain [hunk 2]
```diff
@@ -0,0 +1 @@
+This release introduces a mandatory `federationDomain` configuration setting to cargohold. Please update your `values/wire-server/values.yaml` to set `cargohold.settings.federationDomain` to the same value as the corresponding option in galley (and brig).
```
Comment (fisx) [inline changelog.d/0-release-notes/cargohold-domain:1]: this doesn't match the title of the PR.

File: libs/galley-types/src/Galley/Types/Conversations/Intra.hs [hunk 3]
```diff
@@ -72,7 +72,7 @@ instance ToSchema UpsertOne2OneConversationRequest where
         <*> (qUntagged . uooRemoteUser) .= field "remote_user" (qTagUnsafe <$> schema)
         <*> uooActor .= field "actor" schema
         <*> uooActorDesiredMembership .= field "actor_desired_membership" schema
-        <*> uooConvId .= field "conversation_id" (optWithDefault A.Null schema)
+        <*> uooConvId .= optField "conversation_id" (maybeWithDefault A.Null schema)
```
Comment (fisx) [inline libs/galley-types/src/Galley/Types/Conversations/Intra.hs:75]: how is this related to the PR?

**Review by fisx — COMMENTED** [review]
Could you rebase this so it doesn't contain other PRs any more?

**Comment by pcapriotti** [comment]
> Could you rebase this so it doesn't contain other PRs any more?

Done. The unrelated schema changes are now gone.

File: changelog.d/0-release-notes/cargohold-domain (hunk 2 above)
Comment (pcapriotti) [reply changelog.d/0-release-notes/cargohold-domain:1]: True, but it would be somewhat painful to split this into two PR, since the ZAuth servant combinators have integration with the `Qualified` stuff that needs knowledge of the federation domain, which we'll need in cargohold later anyway. It would just be more pointless work to do this in two steps.

Can I just expand the PR description to include this?

Comment (pcapriotti) [reply changelog.d/0-release-notes/cargohold-domain:1]: I updated the title and description. Ok to merge?
